DataDog/dd-trace-dotnet
# Add workaround for ASP.NET ConfigBuilder issue

## Summary of changes

Adds a check that IisPreStartInit has completed before we run any automatic instrumentations

## Reason for change

We recently had a case where a customer was [using `AzureKeyVaultConfigBuilder`](https://github.com/aspnet/MicrosoftConfigurationBuilders/blob/main/src/Azure/AzureKeyVaultConfigBuilder.cs) with ASP.NET's web.config to load configuration into `AppSettings`, which was causing the  application to deadlock on startup.

After some investigation, we isolated the problem as happening specifically when there are 2 apps running inside an app pool:

- App 1 starts
  - The config builder populates and loads from key vault
  - We initialize the tracer, and set up instrumentation
  - App1 works fine 👍 
- App 2 starts
  - The config builder tries to populate and load from key vault
    - This requires making HTTP requests
    - Due to the instrumentation run for app 1, we instrument the HTTP request
    - This invokes `CallTargetInvoker` which tries to initialize the tracer
    - Initializing the tracer [requires reading `AppSettings` so that we can populate configuration](https://github.com/DataDog/dd-trace-dotnet/blob/c8399df44468f83c22a978418e0cfc314f9c3542/tracer/src/Datadog.Trace/Configuration/ConfigurationSources/GlobalConfigurationSource.cs#L39)
      - The config builder tries to populate and load from key vault 
      - ♻️re-entry 💥

![boom](https://github.com/user-attachments/assets/725ec7ba-a1ff-4069-aa15-f5dd3d28c314)

So the key thing is that we need to make sure we _don't_ run our automatic instrumentation until _after_ the IIS pre-init stage is completely, to avoid re-entry and recursion during setup.

## Implementation details

The implementation is leveraging work we did years ago to fix essentially the same problem: https://github.com/DataDog/dd-trace-dotnet/pull/1157. The problem back then was with Liblog and NLog, but we did all the work to inject flags for tracking when it is safe to make changes.

Given the native work already exists, we can piggy-pack on those hooks and make sure we don't run any automatic instrumentations unless the domain as finished pre-initialization.

This is still _relatively_ fragile, as things like adding a static reference to IDatadogLogger would cause a static initialization to happen too early, and ultimately deadlock/crash the app, so added a bunch of comments to try to highlight the issue

## Test coverage

This is kind of a pain to test because it requires a custom config builder, plus two applications running in an app pool. I tested manually by

- Created a custom `ConfigBuilder` based on the `AzureKeyVaultConfigBuilder` implementation. The custom builder simply makes a generic HTTP request when a value is requested
- Create a generic .NET 472 asp.net app
- Add to the `web.config` to set up the builder (see below)
- Publish the app
- Create two asp.net apps in IIS, using the same app pool
- Hit the first app - all good 👍 
- Hit the second app - 💥 `The configBuilder 'CustomBuilder' failed while processing the configuration section 'appSettings'.: The ConfigurationBuilder 'CustomBuilder[Microsoft.Configuration.ConfigurationBuilders.CustomConfigBuilder]' has recursively re-entered processing of the 'appSettings' section` (I based my implementation on v3 which specifically detects re-entry)
- Made the fix, now it works 🎉 

web.config for my dummy test:
```xml
<configuration>
  <configSections>
    <section name="configBuilders"
      type="System.Configuration.ConfigurationBuildersSection, 
      System.Configuration, Version=4.0.0.0, Culture=neutral, 
      PublicKeyToken=b03f5f7f11d50a3a"
      restartOnExternalChanges="false" requirePermission="false" />
  </configSections>

  <configBuilders>
    <builders>
      <add name="CustomBuilder" preloadSecretNames="false" Uri="https://some-value-prd.vault.azure.net/"
      type="Microsoft.Configuration.ConfigurationBuilders.CustomConfigBuilder, CustomConfigBuilder, Version=1.0.0.0, Culture=neutral" />
    </builders>
  </configBuilders>

  <appSettings configBuilders="CustomBuilder">
    <add key="DummyKey1" value="DummyKey1 value from web.config" />
    <add key="DummyKey2" value="DummyKey2 value from web.config" />
  </appSettings>
```

Complete test solution is here:
[ConfigBuilderIssueRepro.zip](https://github.com/user-attachments/files/17343990/ConfigBuilderIssueRepro.zip)

## Other details

Fixes: https://datadoghq.atlassian.net/browse/APMS-13426
